### DISCUSSION TITLE
Admin API: projects, api keys, user management

### DISCUSSION THREAD

**Discussion by maxxte** [langfuse/langfuse#1007] [Ideas]
### Describe the feature you'd like to request

While automatically setting different environments we would like to add observability in automatic way, this means to create project in langfuse and get secret and public keys, to pass them later to application

### Describe the solution you'd like to see

Introduce new [Projects](https://langfuse.com/docs/api-reference#tag/Projects) endpoint POST with parameters (project name) to create new project
And endpoint for creating API keys, which will consume project name/id and respond with secret and public keys.

### Contribute

- [ ] Yes, I can implement this and raise a PR

**Comment by marcklingen** [comment]
Thanks for raising this, I understand the need. Do you suggest to introduce user-level api keys that can be used to create projects and project-level api keys?

**Reply by IvanAntipovAtExness** [reply]
Yes exacly. 
+add users to the project.

So my pipeline will
- create a langfuse project
- add user to the project
- execute user's LLM queries against newly created project  using API

**Comment by marcklingen** [comment]
@maxxte

**Comment by maxxte** [comment]
@marcklingen, yes, I see this as the only way to control the possibility of creating projects, thank you and sorry for longer response

**Comment by marcklingen** [comment]
Thanks for the clarification! Does a default project/user solve the problem for you or do you need to create multiple projects while and api keys while the container is already running? @maxxte

**Reply by Harryalways317** [reply]
Default project should be fine

**Reply by IvanAntipovAtExness** [reply]
@marcklingen are you suggesting running a new langfuse container per project?

In my scenario, it is not feasible. I will have dozens of users. Each will create a new environment (langfuse project) every day. And I need the tracess to be available at least a month.

So, if I get you right, creating a container per project is a really heavy solution ((

**Reply by marcklingen** [reply]
> So, if I get you right, creating a container per project is a really heavy solution ((

agree, thanks for the feedback

**Comment by mprudra** [comment]
This looks promising.
I want to run a central instance of Langfuse, and different devs can have multiple projects. Currently each user has to create a project, get api keys, set envs, pass callbacks. I want to make this as minimal as possible, we already have a concept of project internally to track LLM usage, may be we can use that to create corresponding Langfuse project for users, set envs and pass callback (we'll see about this, I've some way for this as well, not yet documented by Langchain, @marcklingen we had discussed this earlier in chat, by using Langchain Tracer).

**Comment by mprudra** [comment]
Adding to this, it's best to support create, delete and manage (manage authorization, etc.) APIs.

**Comment by Laurian** [comment]
Additionally to creating projects/keys, which I guess will be authorized via let's say `MASTER_API_KEY`, it would be great to be able to have a default admin user, think of having `NEXT_PUBLIC_SIGN_UP_DISABLED=false` and allowing sign in with `admin` and the `MASTER_API_KEY` as password.

**Comment by hburrichter** [comment]
Adding to this conversation, how about a master API key that can be used for all Langfuse projects? This may already solve the issue of not being able to create projects/keys via an API, for some users, including us.

Currently, we need to manually create each Langfuse project, generate the associated API key, copy them to our secret management tool and have extra business logic in our infrastructure and deployment scripts, that assigns the correct Langchain key to each containerized application.

All this would be instantly solved if we could use one master API key.

**Comment by R4Y-R4Y** [comment]
There's a PR that seeds the database with default parameters and an initial project that has not been looked for https://github.com/langfuse/langfuse/pull/1109

This may help to solve the problem of creating the first project manually but still has the problem of not giving full control to the user in the future. Although i would like the seed PR to be merged since it is what i need to use for docker compose in production

**Comment by maxdeichmann** [comment]
Hi all,

we are currently working on `organizations` where one organization can contain multiple projects. With this change, you will also be able to give users different permissions across projects. Once we have this live, we will also dive into organization level API keys, which would be helpful for the use cases described above.
https://github.com/langfuse/langfuse/pull/1109 is a great example on how we could instantiate an organization API key and allow to create projects with it.

**Comment by danielkerwin** [comment]
This is a big one for us since we would have hundreds of projects (our customer brands), each with hundreds of users per project (each customers users) - it would ideally all roll up to our organization for unified reporting, but each customer will only be able to interact with their project data.

Thus its not feasible to manually create projects, or api keys for our use-case and would definitely need some higher level organization API-based controls

**Comment by marcklingen** [comment]
More context from another thread (#3243) which I would like to merge into this thread to consolidate all discussion on a future admin api

@co3k, @EricAntidot, @thiagolealassis please add your thoughts/upvotes to this thread

## Thread description

by @EricAntidot

> We are self hosting a couple of langfuse instance in our infrastructure and we want to be able to scripts (or integrate with other app) the management part of langfuse project.
> 
> Namely I want to be able to
> 
> - create / delete / update an organization
> - create / delete / update a project
> - create / delete / update a user
> - create / delete / update an API Key
> 
> i.e. we currently do that using langfuse UI but it would be more efficient to do it programmatically.
> 
> The recent added capability to do part of this during init: https://langfuse.com/changelog/2024-09-04-headless-initialization-of-self-hosted-deployments is a step forward. But we want more ;-)


## Implementation idea

> @marcklingen
> We plan to add a proper admin api to manage both
> - organizations
>   - via organization-level api keys (to be introduced)
>   - Example actions: CRUD projects, add/remove members, creates/delete api keys)
> - langfuse self-hosted instances
>   - `ADMIN_API_KEY` to be set via environment
>   - Example actions: CRUD orgs, create/delete org api keys

## Timeline

> > @EricAntidot
>  > Do you have any timeline on those feature?
>  > I'll be glad to review of contribute to the design of this API.
> 
> @marcklingen
> Not yet as there a couple of high-prio changes that we will address beforehand. I think scope/design of apis is quite clear. Please share any additional use cases in addition to the ones mentioned above that you'd need.

## Blockers

> > @co3k
> > For the initial implementation of the organization-level API key issuance functionality that you are envisioning, would it be reasonable to assume that it should be roughly equivalent to the current project-level API key issuance functionality?
> > If so, would it be alright if I propose an implementation via a pull request?
> 
> @marcklingen
> I think this needs a larger revamp of api keys as it would be nice to also use these organization level api keys to perform project-level actions (e.g. use a single api key to add traces to multiple projects, to be differentiated at run-time).

> > @co3k
> > Additionally, considering compatibility with the SDK, I believe it would be best to distinguish between project-level and organization-level API keys solely through their prefixes (for instance, project-level API keys currently use the "pk-" prefix, so organization-level API keys could use the "ok-" prefix). How do you envision distinguishing between these key types?
> 
> @marcklingen
> Good question. I think the ideal scenario here is to only have org-level api keys with custom scopes and migrate the existing keys to have scopes for the single project only. Otherwise this change would add a lot of complexity to the project that is not easy to understand/maintain.

**Comment by marcklingen** [comment]
For user/role/identity management, I'd consider implementing SCIM as part of the admin API instead of creating a custom API. Tracking this here as interest in SCIM might be unrelated to this Admin API: https://github.com/orgs/langfuse/discussions/3546

**Comment by 252afh** [comment]
Hi, I've got langfuse set up self-hosted in AWS. I'm looking to manage our projects and keys after container start-up via terraform, or failing that via an admin API. Has there been any progress on implementing either an admin API or terraform provider? Thanks!

**Reply by marcklingen** [reply]
Not yet, thanks on the +1 on this!

**Comment by theabhinavdas** [comment]
@marcklingen if the team hasn't begun implementation of this yet, can i pick this up?

**Comment by marcklingen** [comment]
@adedj in https://github.com/orgs/langfuse/discussions/6105

> We would like to have APIs that will allow us to:
> 
> manage users (invite / delete )
> manage organizations (create /delete, add/remove members, change user role)
> manage projects (create, delete, add/remove members, change user role, transfer project, configure future integrations)
> This will help us manage the langfuse platform more easily in an automated way.

**Comment by slyt** [comment]
We're looking at deploying Langfuse to monitor no-code AI agent deployments. 

As __Maintainers__* (the people deploying Langfuse), we'd like to _programmatically_:
- Create Langfuse Organizations and add members that own the agent
- Create a Langfuse Project that is one-to-one mapping with the agent
- Retrieve the API keys for the project and then inject them into the agent so that Langfuse can monitor it

We'd also like a "Maintainers" group to be able to globally view all Projects in all Organization in the Langfuse instance. It does not seem like the current [RBAC](https://langfuse.com/docs/rbac) system would provide this functionality.

`* I use the term "Maintainers" to avoid confusion with the "Admin" role in RBAC.`

**Reply by marcklingen** [reply]
thanks for the additional context here!

**Comment by slyt** [comment]
@marcklingen I'm interested in contributing this feature. 

Before I dive in, I’d love your guidance on a few things:
- Are there any known blockers, refactors, or key decisions still pending that I should be aware of?
- Is there a recommended starting point (branch, issue, or design doc)?
- At this stage, would contributions be more helpful around design, implementation, or feedback?

Happy to collaborate and iterate in whichever way is most useful.

**Comment by rehevkor5** [comment]
> Adding to this conversation, how about a master API key that can be used for all Langfuse projects? This may already solve the issue of not being able to create projects/keys via an API, for some users, including us.

My company would find that helpful. We have a few centralized API services which our chat bots/agents are either hosted by or make API calls to. We want those centralized services to be able to produce traces to Langfuse, however it is awkward that we have to ask people to create a Langfuse API key for their Project and then give it to (configure it onto) our services so they can produce traces.

Ideally, we could generate a Langfuse API key for each of those centralized services. Each of those API keys would have permission to produce traces into Langfuse under any Project.

**Comment by marcklingen** [comment]
This has been available for a couple of months, but I missed the opportunity to update this thread sooner. Sorry!

APIs
- Instance level: https://langfuse.com/self-hosting/administration/organization-management-api
- Org-level https://langfuse.com/docs/administration/scim-and-org-api

**Comment by kinthaiofficial** [comment]
For multi-agent platforms specifically, the Admin API needs a few things that go beyond standard project/API key management:

**Agent-scoped keys** — you want API keys that are scoped to a specific agent identity, not just a project. When agent A delegates to agent B, B's calls should be attributable to B's key within A's project budget, not just to "project X."

**Hierarchical quota enforcement** — project-level rate limits are coarse. What you really need is: namespace → user → agent → conversation, each with its own ceiling. A runaway agent in user A's context shouldn't be able to exhaust the project quota for user B.

**Capability manifests on keys** — which API endpoints this key is allowed to call. An agent key for a retrieval-only agent shouldn't be able to call the generation endpoint. Monotonic narrowing when you create child keys from parent keys.

**Audit log per agent** — not just "who called the API" but "which agent in which delegation chain called the API at what cost."

We built a version of this for KinthAI's agent identity layer — the tricky part was making the key hierarchy enforceable without adding latency to every API call. Write-up here: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons covers the identity/delegation design.

Is the Admin API use case primarily for SaaS customers managing their own teams, or for programmatic agent fleet management?
